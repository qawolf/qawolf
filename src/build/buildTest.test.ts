import { loadEvents } from "@qawolf/fixtures";
import { buildJob } from "./buildJob";
import { buildTest, formatStep } from "./buildTest";

describe("buildTest", () => {
  test("builds a test from a job", async () => {
    const events = await loadEvents("login");
    const job = buildJob(events, "test job");
    const testString = buildTest(job, false);

    expect(testString).toMatchSnapshot();
  });
});

describe("formatStep", () => {
  test("formats click on link step", () => {
    const step = {
      action: "click" as "click",
      target: {
        tagName: "a",
        textContent: "contact"
      }
    };

    const formattedStep = formatStep(step);
    expect(formattedStep).toEqual("click contact link");
  });

  test("formats click on submit input step", () => {
    const step = {
      action: "click" as "click",
      target: {
        inputType: "submit",
        tagName: "input",
        textContent: "sign in"
      }
    };

    const formattedStep = formatStep(step);
    expect(formattedStep).toEqual('click sign in input[type="submit"]');
  });

  test("formats scroll down action", () => {
    const step = {
      action: "scroll" as "scroll",
      target: { xpath: "scroll" },
      scrollDirection: "down" as "down"
    };

    const formattedStep = formatStep(step);
    expect(formattedStep).toEqual("scroll down");
  });

  test("formats scroll up action", () => {
    const step = {
      action: "scroll" as "scroll",
      target: { xpath: "scroll" },
      scrollDirection: "up" as "up"
    };

    const formattedStep = formatStep(step);
    expect(formattedStep).toEqual("scroll up");
  });

  test("formats type into text input", () => {
    const step = {
      action: "input" as "input",
      target: {
        id: "input1",
        inputType: "text",
        labels: ["username"],
        name: "user",
        placeholder: "Jane Doe",
        tagName: "input"
      },
      value: "spirit"
    };

    const formattedStep = formatStep(step);
    expect(formattedStep).toEqual(
      'enter "spirit" into username input[type="text"]'
    );
  });

  test("formats type into password input", () => {
    const step = {
      action: "input" as "input",
      target: {
        id: "input2",
        inputType: "password",
        placeholder: "secret",
        tagName: "input"
      },
      value: "supersecret"
    };

    const formattedStep = formatStep(step);
    expect(formattedStep).toEqual(
      'enter "supersecret" into secret input[type="password"]'
    );
  });

  test("shortens target name if needed", () => {
    const step = {
      action: "click" as "click",
      target: {
        inputType: "submit",
        tagName: "input",
        textContent: `sign in${"x".repeat(200)}`
      }
    };

    const formattedTextContent = `sign in${"x".repeat(43)}`;

    const formattedStep = formatStep(step);
    expect(formattedStep).toEqual(
      `click ${formattedTextContent} input[type="submit"]`
    );
  });
});
