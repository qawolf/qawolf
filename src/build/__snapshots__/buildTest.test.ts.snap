// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildTest builds a test from a job 1`] = `
"const { Browser, hooks } = require(\\"qawolf\\");

let browser;

beforeAll(async () => {
  browser = await Browser.create({ size: \\"desktop\\", url: \\"http://localhost:5000/\\" });
  await hooks.beforeAll(browser);
});

beforeEach(() => hooks.beforeEach(browser));

afterEach(() => hooks.afterEach(browser));

afterAll(() => hooks.afterAll(browser));

test('scroll down', async () => {
  await browser.runStep({\\"action\\":\\"scroll\\",\\"target\\":{\\"xpath\\":\\"scroll\\"},\\"pageId\\":0,\\"scrollDirection\\":\\"down\\",\\"scrollTo\\":334});
});

test('click form authentication link', async () => {
  await browser.runStep({\\"action\\":\\"click\\",\\"target\\":{\\"classList\\":null,\\"href\\":\\"http://localhost:5000/login\\",\\"id\\":null,\\"inputType\\":null,\\"labels\\":null,\\"name\\":null,\\"parentText\\":[\\"form authentication\\",\\"form authentication\\"],\\"placeholder\\":null,\\"tagName\\":\\"a\\",\\"textContent\\":\\"form authentication\\",\\"xpath\\":\\"//*[@id='content']/ul/li[18]/a\\"},\\"pageId\\":0});
});

test('enter \\"tomsmith\\" into username input[type=\\"text\\"]', async () => {
  await browser.runStep({\\"action\\":\\"input\\",\\"target\\":{\\"classList\\":null,\\"href\\":null,\\"id\\":\\"username\\",\\"inputType\\":\\"text\\",\\"labels\\":[\\"username\\"],\\"name\\":\\"username\\",\\"parentText\\":[\\"username\\",\\"\\\\n        username\\\\n        \\\\n      \\"],\\"placeholder\\":null,\\"tagName\\":\\"input\\",\\"textContent\\":null,\\"xpath\\":\\"//*[@id='username']\\"},\\"pageId\\":0,\\"value\\":\\"tomsmith\\"});
});

test('click login button[type=\\"submit\\"]', async () => {
  await browser.runStep({\\"action\\":\\"click\\",\\"target\\":{\\"classList\\":[\\"radius\\"],\\"href\\":null,\\"id\\":null,\\"inputType\\":\\"submit\\",\\"labels\\":null,\\"name\\":null,\\"parentText\\":[\\"\\\\n      \\\\n        username\\\\n        \\\\n      \\\\n    \\",\\"\\\\n      \\\\n        password\\\\n        \\\\n      \\\\n    \\",\\"login\\",\\"\\\\n     \\\\n      \\\\n        username\\\\n        \\\\n      \\\\n    \\\\n    \\\\n      \\\\n        password\\\\n        \\\\n      \\\\n    \\\\n       login\\\\n  \\"],\\"placeholder\\":null,\\"tagName\\":\\"button\\",\\"textContent\\":\\"login\\",\\"xpath\\":\\"//*[@id='login']/button\\"},\\"pageId\\":0});
});

test('enter \\"tomsmith\\" into username input[type=\\"text\\"]', async () => {
  await browser.runStep({\\"action\\":\\"input\\",\\"target\\":{\\"classList\\":null,\\"href\\":null,\\"id\\":\\"username\\",\\"inputType\\":\\"text\\",\\"labels\\":[\\"username\\"],\\"name\\":\\"username\\",\\"parentText\\":[\\"username\\",\\"\\\\n        username\\\\n        \\\\n      \\"],\\"placeholder\\":null,\\"tagName\\":\\"input\\",\\"textContent\\":null,\\"xpath\\":\\"//*[@id='username']\\"},\\"pageId\\":0,\\"value\\":\\"tomsmith\\"});
});

test('enter \\"SuperSecretPassword!\\" into password input[type=\\"password\\"]', async () => {
  await browser.runStep({\\"action\\":\\"input\\",\\"target\\":{\\"classList\\":null,\\"href\\":null,\\"id\\":\\"password\\",\\"inputType\\":\\"password\\",\\"labels\\":[\\"password\\"],\\"name\\":\\"password\\",\\"parentText\\":[\\"password\\",\\"\\\\n        password\\\\n        \\\\n      \\"],\\"placeholder\\":null,\\"tagName\\":\\"input\\",\\"textContent\\":null,\\"xpath\\":\\"//*[@id='password']\\"},\\"pageId\\":0,\\"value\\":\\"SuperSecretPassword!\\"});
});

test('click login i', async () => {
  await browser.runStep({\\"action\\":\\"click\\",\\"target\\":{\\"classList\\":[\\"fa\\",\\"fa-2x\\",\\"fa-sign-in\\"],\\"href\\":null,\\"id\\":null,\\"inputType\\":null,\\"labels\\":null,\\"name\\":null,\\"parentText\\":[\\"login\\",\\"login\\"],\\"placeholder\\":null,\\"tagName\\":\\"i\\",\\"textContent\\":\\"login\\",\\"xpath\\":\\"//*[@id='login']/button/i\\"},\\"pageId\\":0});
});
"
`;
